### PR TITLE
[PostgreSQL] Support forced SSL-verified connection with given CA certificate

### PR DESCRIPTION
**Pull request by veger**
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
postgresql modules

##### ANSIBLE VERSION
```
ansible 2.2.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY
Add support for a forced SSL connection to a PostgreSQL databases using a given certificate.
When the certificate does not match (or is incorrect at any way), the connection is refused.

### CODE REVIEW

**Comment by veger** [comment]
Migrated from https://github.com/ansible/ansible-modules-core/pull/5698

File: lib/ansible/modules/database/postgresql/postgresql_privs.py
Comment (gundalow) [inline lib/ansible/modules/database/postgresql/postgresql_privs.py]: Missing full stop at the end of this sentence, and the one later on

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/database/postgresql/postgresql_privs.py
Comment (gundalow) [inline lib/ansible/modules/database/postgresql/postgresql_privs.py:123]: Missing example or detail of what the valid options are.

**Review by gundalow — CHANGES_REQUESTED** [review]
needs_revision 

Docs/examples need expanding to explain in more details how to use the options.

**Comment by veger** [comment]
Implemented feedback of @gundalow.

Additionally, updated the description of `ssl_mode` with the possible choices and added 'forced' default instead of relying on the PostgreSQL default implementation (which might change in the future...)

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
Thanks for the updates.

I'll let someone else from the Core Team look at this then they can merge into `devel`

**Comment by alikins** [comment]
Should the new modules options be sslmode/sslrootcert (to match the postgresql names?)

**Comment by veger** [comment]
I matched the modules parameter style/names, which seemed most user-friendly to me.

If the (generic) preference is to support the postgresql names, let me know and I'll make the changes.

File: lib/ansible/modules/database/postgresql/postgresql_privs.py
Comment (abadger) [inline lib/ansible/modules/database/postgresql/postgresql_privs.py]: default here is empty string but both in the documentation and in the other file, postgresql_user.py , it's None.  Should standardize on either None or empty string in all places.

**Comment by abadger** [comment]
Left some comments.  On the question of how to spell the module parameters, I can see the argument either way.  I think matching the other module params is probably the slightly better way to go.

Once the comments are addressed this can be merged.

**Comment by abadger** [comment]
needs_revision

File: lib/ansible/modules/database/postgresql/postgresql_privs.py
Comment (veger) [reply lib/ansible/modules/database/postgresql/postgresql_privs.py]: I tried to keep the same format per module. Each module uses its own format (empty string/ None/ etc). I agree that standardizing this is the way to go! But I also think that this is outside the scope of this PR..?

Comment (abadger) [reply lib/ansible/modules/database/postgresql/postgresql_privs.py]: Not sure I understand what you're saying?  This PR adds ssl_rootcert to both modules.  Seems like they should both default to either empty string or both default to None?

Comment (veger) [reply lib/ansible/modules/database/postgresql/postgresql_privs.py]: postgres_db.py uses `""` for its defaults, postgres_privs.py uses `''` and postgres_user.py uses `None`. This PR tries to adhere to the format/style of each file (which differs between each other...).

Postgres itself does not mind either way.

Comment (abadger) [reply lib/ansible/modules/database/postgresql/postgresql_privs.py]: If postgres doesn't care, let's standardize on None in all three modules.

**Comment by abadger** [comment]
Only one thing to change.  Once that's done, this is good to merge.

**Comment by abadger** [comment]
@veger do you have everything you need to update this or do you need more information from me?

**Comment by veger** [comment]
I have everything (but time :P ), hopefully I'll get to it tomorrow

**Comment by veger** [comment]
Changed the defaults to `None`

**Comment by abadger** [comment]
/me trying to figure out why the postgres integration tests failed for this PR.

**Comment by abadger** [comment]
Ahh... you misunderstood me. Don't mess with the defaults for the other options.  Only change the defaults for the ssl options you're adding.  Changing the other options breaks the code.

**Comment by veger** [comment]
Oops... I have reverted those changes. Hopefully the tests will pass now!

**Comment by veger** [comment]
Hm... looks like the PostgreSQL installed on CentOS 6.1 does not support the `sslrootcert` parameter.

Can we ignore this, or is this a serious issue?

**Comment by abadger** [comment]
@veger it is an issue but we can conditionalize use of the parameter on the version of the server.  In the documentation, update it to say that the sslrootcert parameter is only valid for postgresql server >= X.Y

In the code it'll be trickier because we need the parameter in order to connect.  I suggest that we omit the parameter if it's None (since that's the default and means the user hasn't set it).  If it's not None and the connection to the server fails, it will be nice to look at the error message and if it tells us that it fails because of sslrootcert, then we return a nice error via module.fail_json() telling the user that postgresql needs to be at least version X.Y in order to use sslrootcert.

**Comment by abadger** [comment]
@vger okay, I haven't tested but I think this is about what you need to do.  I only changed postgres_db but the other files should be similar: https://toshio.fedorapeople.org/ansible/postgres.patch

**Comment by veger** [comment]
@abadger Thanks for the patch, I used it as a basis to modify the 3 modules.

Centos 6.1 tests now passing as well

**Comment by abadger** [comment]
Thanks @veger!  Merged to the devel branch for inclusion in 2.3.0

**Pull request merged**
